sync2/multipeer/multipeer_test.go

```diff
@@ -0,0 +1,376 @@
+				FP:    "foo",
+				Count: 100,
+				Sim:   0.99, // high enough for full sync
+			})
+			mt.expectFullSync(pl, 6, 0)
+			mt.syncBase.EXPECT().Wait()
+			if i == 0 {
+				//nolint:fatcontext
+				ctx = mt.start()
+			} else {
+				// first full sync happens immediately
+				mt.clock.Advance(time.Minute)
+			}
+			mt.clock.BlockUntilContext(ctx, 1)
+			mt.satisfy()
+		}
+		require.True(t, mt.reconciler.Synced())
+		mt.syncBase.EXPECT().Wait()
+	})
+
+	t.Run("full sync, peers with low count ignored", func(t *testing.T) {
+		mt := newMultiPeerSyncTester(t)
+		addedPeers := mt.addPeers(6)
+		mt.syncBase.EXPECT().Count().Return(1000, nil).AnyTimes()
+		require.False(t, mt.reconciler.Synced())
+		var ctx context.Context
+		for i := 0; i < numSyncs; i++ {
+			var pl peerList
+			for _, p := range addedPeers[:5] {
+				mt.expectSingleProbe(p, rangesync.ProbeResult{
+					FP:    "foo",
+					Count: 1000,
+					Sim:   0.99, // high enough for full sync
+				})
+				pl.add(p)
+			}
+			mt.expectSingleProbe(addedPeers[5], rangesync.ProbeResult{
+				FP:    "foo",
+				Count: 800, // count too low, this peer should be ignored
+				Sim:   0.9,
+			})
+			mt.expectFullSync(&pl, 5, 0)
+			mt.syncBase.EXPECT().Wait()
+			if i == 0 {
+				//nolint:fatcontext
+				ctx = mt.start()
+			} else {
+				// first full sync happens immediately
+				mt.clock.Advance(time.Minute)
+			}
+			mt.clock.BlockUntilContext(ctx, 1)
+			mt.satisfy()
+		}
+		require.True(t, mt.reconciler.Synced())
+		mt.syncBase.EXPECT().Wait()
+	})
+
+	t.Run("full sync due to low peer count", func(t *testing.T) {
+		mt := newMultiPeerSyncTester(t)
+		mt.addPeers(1)
+		mt.syncBase.EXPECT().Count().Return(50, nil).AnyTimes()
+		var ctx context.Context
+		for i := 0; i < numSyncs; i++ {
+			pl := mt.expectProbe(1, rangesync.ProbeResult{
+				FP:    "foo",
+				Count: 100,
+				Sim:   0.5, // too low for full sync, but will have it anyway
+			})
+			mt.expectFullSync(pl, 1, 0)
+			mt.syncBase.EXPECT().Wait()
+			if i == 0 {
+				//nolint:fatcontext
+				ctx = mt.start()
+			} else {
+				mt.clock.Advance(time.Minute)
+			}
+			mt.clock.BlockUntilContext(ctx, 1)
+			mt.satisfy()
+		}
+		mt.syncBase.EXPECT().Wait()
+	})
+
+	t.Run("probe failure", func(t *testing.T) {
+		mt := newMultiPeerSyncTester(t)
+		mt.addPeers(10)
+		mt.syncBase.EXPECT().Count().Return(100, nil).AnyTimes()
+		mt.syncBase.EXPECT().Probe(gomock.Any(), gomock.Any()).
+			Return(rangesync.ProbeResult{}, errors.New("probe failed"))
+		pl := mt.expectProbe(5, rangesync.ProbeResult{FP: "foo", Count: 100, Sim: 0.99})
+		// just 5 peers for which the probe worked will be checked
+		mt.expectFullSync(pl, 5, 0)
+		mt.syncBase.EXPECT().Wait().Times(2)
+		ctx := mt.start()
+		mt.clock.BlockUntilContext(ctx, 1)
+	})
+
+	t.Run("failed peers during full sync", func(t *testing.T) {
+		mt := newMultiPeerSyncTester(t)
+		mt.addPeers(10)
+		mt.syncBase.EXPECT().Count().Return(100, nil).AnyTimes()
+		var ctx context.Context
```

Comment:
`mt.start()` creates the actual context and sets it to be canceled when the text finishes, after which it is checked that `reconciler.Run()` has returned `context.Canceled` and not some other unexpected error.
Splitting `mt.start()` to create the context separately is somewhat inconvenient.
But `mt.start()` cannot be called before we've set the expectations, otherwise we get a race in the tests.